### DISCUSSION TITLE
Best way to do source code modifications?

### DISCUSSION THREAD

**Discussion by chrisjsewell** [aiidateam/aiida-upgrade#1] [General]
Some examples:

- https://github.com/pytest-dev/unittest2pytest
- https://github.com/asottile/pyupgrade
- https://github.com/PyCQA/isort

**Comment by ltalirz** [comment]
Just mentioning that we wrote something to help with migration of plugins for the 0.x to 1.x transition that was using bowler https://github.com/aiidateam/aiida-plugin-migrator

**Reply by chrisjsewell** [reply]
cheers

**Comment by chrisjsewell** [comment]
Was looking at https://stackoverflow.com/q/768634/5033292, and note in python 3.9 there is now https://docs.python.org/3.9/library/ast.html#ast.unparse which could be useful

**Reply by chrisjsewell** [reply]
and from that, https://github.com/isidentical/refactor also looks interesting

**Reply by mbercx** [reply]
Seems pure AST is lossy because it doesn't preserve whitespace/comments?

`refactor` looks interesting but I can't pip install it for some reason 🤷 

Now looking at https://github.com/Instagram/LibCST/

**Reply by ltalirz** [reply]
and what we used for the previous plugin migrator is not powerful enough?
https://github.com/facebookincubator/Bowler

**Reply by mbercx** [reply]
Right, I'll have a look at that as well 👍

**Reply by mbercx** [reply]
In the end I decided to use LibCST over bowler. Although bowler is a nice tool building on other CST packages that provides a very easy API for simple changes, [their development seems to be stalled atm](https://github.com/facebookincubator/Bowler#current-development-status) (they actually recommend using LibCST), so I figured I'd go straight for the more powerful tool. In the end I was interesting in learning about AST/CST anyways, and was more quickly able to figure out how to e.g.  add the `core.` prefix for factories in LibSCT.
